pkgs/extension_discovery/lib/src/package_config.dart

```diff
@@ -8,6 +8,36 @@ import 'expect_json.dart';
 typedef PackageConfigEntry = ({String name, Uri rootUri, Uri packageUri});
 typedef PackageConfig = List<PackageConfigEntry>;
 
+/// Searches in [packageDir] and all directories above for a
+/// `.dart_tool/package_config.json` file. Returns the Uri of that file if
+/// found.
+///
+/// Returns `null` if no file was found.
+Uri? findPackageConfig(Uri packageDir) {
+  if (!packageDir.isScheme('file')) {
+    throw ArgumentError(
+      'Expected [packageDir] to be a file URI, got $packageDir',
+    );
+  }
+  if (!packageDir.path.endsWith('/')) {
+    packageDir = packageDir.replace(path: '${packageDir.path}/');
+  }
+  while (true) {
+    final packageConfigCandidate =
+        packageDir.resolve('.dart_tool/package_config.json');
```

Comment:
Yep, this looks reasonable to me.

The things that usually cause issues are reading `.path` from a File URI, assuming it's correct for Windows (it's not, but it is correct for non-Windows), and going to/from Windows paths when running in the browser (because Dart always treats web as non-Windows even if the `file:///` URI is a Windows path).

I note the code below is doing `File.fromUri()` which will call `toFilePath()` but since it's calling `existsSync()` I assume that this code only runs on the host machine (not in a browser) and therefore I don't think any of those gotchas apply here.

Reply:
> I assume that this code only runs on the host machine (not in a browser) 

Yes, this is running on the host.